proxy/threads.c

```diff
@@ -0,0 +1,176 @@
+
+
+
+void* connectionThread(void* arg) {
+    Data d = *(Data*)arg;
+
+    queue_t* newConnectQueue = d.queue;
+    sPoll* spoll = d.poll;
+    //ssize_t bytes_read;
+
+    for (;;) {
+        Sockets* newSockets;
+        queue_get(newConnectQueue, (void**)&newSockets);
+        char buffer[BUFFER_SIZE];
+        // Чтение запроса от клиента
+        int readed;
+        //printf("начал читать %d\n", newSockets->client);
+        if (( readed = read(newSockets->client, buffer, BUFFER_SIZE)) > 0) {
+        }
+
+        if (readed == -1) {
+            printf("соединение разорвано\n");
+            closeSockets(*newSockets);
+            free(newSockets);
+            continue;
+        }
+
+        int dnslen;
+        char *start, *end;
+
+        start = strstr(buffer, "//");
+        if (!start) {
+            closeSockets(*newSockets);
+            free(newSockets);
+            continue;
+        }
+        start+=2;
+
+        end = strchr(start, '/');
+        if (!end) {
+            closeSockets(*newSockets);
+            free(newSockets);
+            continue;     
+        }
+        dnslen = end-start;
+
+        
+
+        char* dns = (char*)malloc(dnslen+1);
+        strncpy(dns, start, dnslen);
+        
+
+
+        const char *port = "80";
+        const char *path = "/";
+        // Получаем IP-адрес по доменному имени
+        struct addrinfo hints, *result, *rp;
+        memset(&hints, 0, sizeof(struct addrinfo));
+        hints.ai_family = AF_UNSPEC;
+        hints.ai_socktype = SOCK_STREAM;
+    
+        int status = getaddrinfo(dns, port, &hints, &result);
+        if (status != 0) {
+            fprintf(stderr, "getaddrinfo: %s\n", gai_strerror(status));
+            close(newSockets->client);
+            continue;;
+        }
+
+        // Перебираем результаты, используем первый подходящий IP-адрес
+        int soc = -1;
+        for (rp = result; rp != NULL; rp = rp->ai_next) {
+            soc = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
+            if (soc == -1) {
+                continue; // Ошибка при создании сокета, пробуем следующий адрес
+            }
+
+            if (connect(soc, rp->ai_addr, rp->ai_addrlen) != -1) {
+                break; // Успешное соединение
+            }
+    
+            close(soc); // Ошибка соединения, закрываем сокет и пробуем следующий адрес
+        }
+
+        freeaddrinfo(result); // Освобождаем структуры addrinfo
+
+        newSockets->server = soc;
+
+        int sended = 0;
+        while ( sended != strlen(buffer)) {
+            int n = send(newSockets->server, buffer, strlen(buffer), 0);
+            if (n == -1) {
+                sended = -1;
+                break;
+            }
+            sended+=n;
+        }
+        if (sended == -1) {
+            closeSockets(*newSockets);
+            free(newSockets);
+            continue;
+        }
+        memset(buffer, 0, BUFFER_SIZE);
+
+        sPollPush(spoll, newSockets);
```

Comment:
you mean you return newSockets into poll queue to restart this cycle again  once more ?  I cannot understand logic of your implementation :( Please explain it briefly.